File: typescript/src/tools.ts
/** Custom tools registered alongside MCP-discovered ones. */
import * as fs from "node:fs";
import * as path from "node:path";
import { tool } from "@langchain/core/tools";
import { z } from "zod";

const IMAGES_DIR = path.resolve(import.meta.dirname, "..", "images");

const EXT_TO_MIME: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
};

function detectMime(buf: Buffer, ext: string): string {
  if (buf[0] === 0x89 && buf.toString("ascii", 1, 4) === "PNG") {
    return "image/png";
  }
  if (buf[0] === 0xff && buf[1] === 0xd8 && buf[2] === 0xff) {
    return "image/jpeg";
  }
  const header6 = buf.toString("ascii", 0, 6);
  if (header6 === "GIF87a" || header6 === "GIF89a") return "image/gif";
  if (
    buf.length >= 12 &&
    buf.toString("ascii", 0, 4) === "RIFF" &&
    buf.toString("ascii", 8, 12) === "WEBP"
  ) {
    return "image/webp";
  }
  return EXT_TO_MIME[ext.toLowerCase()] ?? "image/png";
}

function dataUri(filePath: string): string {
  const buf = fs.readFileSync(filePath);
  const ext = path.extname(filePath);
  const mime = detectMime(buf, ext);
  const b64 = buf.toString("base64");
  return `data:${mime};base64,${b64}`;
}

export const sampleImage = tool(
  async () => {
    if (!fs.existsSync(IMAGES_DIR)) {
      return "No images found in the images/ directory.";
    }
    const files = fs
      .readdirSync(IMAGES_DIR)
      .filter((f) => Object.keys(EXT_TO_MIME).includes(path.extname(f).toLowerCase()))
      .sort();
    if (files.length === 0) {
      return "No images found in the images/ directory.";
    }

    const content: Record<string, unknown>[] = [
      { type: "text", text: "Sample image captured." },
    ];
    for (let i = 0; i < files.length; i++) {
      const uri = dataUri(path.join(IMAGES_DIR, files[i]));
      if (i % 2 === 0) {
        content.push({ type: "image_url", image_url: { url: uri } });
      } else {
        content.push({ type: "input_image", image_url: uri });
      }
      content.push({ type: "text", text: `Image ${i + 1} attached.` });
    }
    return JSON.stringify(content);
  },
  {
    name: "sample_image",
    description:
      "Return a sample image for multimodal testing. Call this tool whenever " +
      "the user asks to see a sample image or wants to test image handling.",
    schema: z.object({}),
  },
);

export function getCustomTools() {
  return [sampleImage];
}

File: python/src/tools.py
"""Custom tools registered alongside MCP-discovered ones."""
import base64
from pathlib import Path

from langchain_core.messages import ToolMessage
from langchain_core.tools import tool

IMAGES_DIR = Path(__file__).resolve().parent.parent / "images"

_EXT_TO_MIME = {
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".webp": "image/webp",
}


def _read_b64(path: Path) -> tuple[str, str]:
    """Read an image file and return (mime, base64_data)."""
    data = path.read_bytes()
    if data[:8] == b'\x89PNG\r\n\x1a\n':
        mime = "image/png"
    elif data[:3] == b'\xff\xd8\xff':
        mime = "image/jpeg"
    elif data[:6] in (b'GIF87a', b'GIF89a'):
        mime = "image/gif"
    elif len(data) >= 12 and data[:4] == b'RIFF' and data[8:12] == b'WEBP':
        mime = "image/webp"
    else:
        mime = _EXT_TO_MIME.get(path.suffix.lower(), "image/png")
    return mime, base64.b64encode(data).decode("ascii")


def _data_uri(path: Path) -> str:
    mime, b64 = _read_b64(path)
    return f"data:{mime};base64,{b64}"


@tool
def sample_image() -> str:
    """Return a sample image for multimodal testing. Call this tool whenever
    the user asks to see a sample image or wants to test image handling."""
    images = sorted(IMAGES_DIR.glob("*"))
    images = [p for p in images if p.suffix.lower() in _EXT_TO_MIME]
    if not images:
        return "No images found in the images/ directory."

    content: list[dict] = [{"type": "text", "text": "Sample image captured."}]
    for i, p in enumerate(images):
        uri = _data_uri(p)
        if i % 2 == 0:
            content.append({
                "type": "image_url",
                "image_url": {"url": uri},
            })
        else:
            content.append({
                "type": "input_image",
                "image_url": uri,
            })
        content.append({"type": "text", "text": f"Image {i + 1} attached."})
    return content


def get_custom_tools() -> list:
    """Return all custom (non-MCP) tools to register with the agent."""
    return [sample_image]
